infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -0,0 +1,339 @@
+
+                List<LayoutStripe> newStripes = layoutSegment.getStripes().stream().map(stripe -> {
+
+                    if (layoutSegment.getFirstStripe().equals(stripe)) {
+                        return new LayoutStripe(
+                                new ImmutableList.Builder<String>()
+                                        .addAll(stripe.getLogServers()).add(server).build());
+                    }
+
+                    return stripe;
+
+                }).collect(Collectors.toList());
+
+                return new LayoutSegment(layoutSegment.getReplicationMode(),
+                        layoutSegment.getStart(), layoutSegment.getEnd(), newStripes);
+            } else {
+                return new LayoutSegment(layoutSegment.getReplicationMode(),
+                        layoutSegment.getStart(), layoutSegment.getEnd(), layoutSegment.getStripes());
+            }
+        }).collect(Collectors.toList());
+        Layout newLayout = new Layout(layout);
+        LayoutBuilder builder = new LayoutBuilder(newLayout);
+        return builder.setSegments(segments).build();
+    }
+
+    /**
+     * Updates the initial layout given the restored transferred segments.
+     *
+     * @param segments   A list of restored transferred segments.
+     * @param initLayout Initial layout.
+     * @return A new, updated layout.
+     */
+    public Layout updateLayoutAfterRedundancyRestoration(
+            List<CurrentTransferSegment> segments, Layout initLayout) {
+        return segments
+                .stream()
+                .reduce(initLayout,
+                        (layout, segment) -> restoreRedundancyForSegment(segment, layout),
+                        (oldLayout, newLayout) -> newLayout);
+    }
+
+    /**
+     * Returns true if a given layout requires redundancy restoration.
+     *
+     * @param layout Current layout.
+     * @param server Current server.
+     * @return True, if requires.
+     */
+    public static boolean requiresRedundancyRestoration(Layout layout, String server) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+            return layout.getSegments().stream().anyMatch(segment -> !segment.getAllLogServers()
+                    .contains(server));
+        }
+    }
+
+    /**
+     * Returns true if after adding a server to the first segment, the segments can be merged.
+     *
+     * @param layout Current layout.
+     * @param server The current node.
+     * @return True is there is a redundancy restoration is needed.
+     */
+    public static boolean requiresMerge(Layout layout, String server) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+            int firstSegmentIndex = 0;
+            int secondSegmentIndex = 1;
+            Layout copy = new Layout(layout);
+            LayoutStripe firstStripe = copy.getSegments().get(0).getFirstStripe();
```

Comment:
if segments is an empty list then it throws an exception.
Please replace by `copy.getSegments().stream().findFirst()`